src/main/scala/is/hail/variant/Call.scala

```diff
@@ -3,113 +3,318 @@ package is.hail.variant
+      case _ => throw new UnsupportedOperationException
+    }
+  }
+}
+
 object Call extends Serializable {
+  def apply(ar: Int, phased: Boolean, ploidy: Int): Call = {
+    require(ploidy >= 0 && ploidy <= 2, s"invalid ploidy: $ploidy. Only support ploidy in range [0, 2]")
+    require(ar >= 0, s"invalid allele representation: $ar. Must be positive.")
+
+    var c = 0
+    c |= phased.toInt
+
+    if (ploidy > 2)
+      c |= (3 << 1)
+    else
+      c |= (ploidy << 1)
 
-  def apply(call: Int): Call = {
-    require(call >= 0, s"Call must be null or >= 0. Found ${ call }.")
-    call
+    require((ar >>> 29) == 0, s"invalid allele representation: $ar. Max value is 2^29 - 1")
+    c |= ar << 3
+
+    c
   }
 
-  def toString(call: Call): String = {
-      val p = Genotype.gtPair(call)
-      s"${ p.j }/${ p.k }"
+  def isPhased(c: Call): Boolean = (c & 0x1) == 1
+
+  def isHaploid(c: Call): Boolean = ploidy(c) == 1
+
+  def isDiploid(c: Call): Boolean = ploidy(c) == 2
+
+  def isUnphasedDiploid(c: Call): Boolean = (c & 0x7) == 4
+
+  def isPhasedDiploid(c: Call): Boolean = (c & 0x7) == 5
+
+  def ploidy(c: Call): Int = (c >>> 1) & 0x3
+
+  def alleleRepr(c: Call): Int = c >>> 3
+
+  def allelePair(c: Call): AllelePair = {
+    require(isDiploid(c), s"invalid ploidy: ${ ploidy(c) }. Only support ploidy == 2")
+
+    if (isPhased(c)) {
+        val p = Genotype.allelePair(alleleRepr(c))
+        AllelePair(p.j, p.k - p.j)
+    } else
+      Genotype.allelePair(alleleRepr(c))
+  }
+
+  def unphasedDiploidGtIndex(c: Call): Int = {
+    require(isUnphasedDiploid(c), s"Only support ploidy == 2 and unphased. Found ${ Call.toString(c) }.")
+    alleleRepr(c)
+  }
+
+  def alleles(c: Call): Array[Int] = {
+    (ploidy(c): @switch) match {
+      case 0 => Array.empty[Int]
+      case 1 => Array(alleleByIndex(c, 0))
+      case 2 => allelePair(c).alleleIndices
+      case _ => throw new UnsupportedOperationException
     }
+  }
 
-  def check(call: Call, nAlleles: Int) {
-    val nGenotypes = triangle(nAlleles)
-    assert(call >= 0 && call < nGenotypes, s"Invalid genotype found `$call' for number of alleles equal to `$nAlleles'.")
+  def alleleByIndex(c: Call, i: Int): Int = {
+    (ploidy(c): @switch) match {
+      case 0 => throw new UnsupportedOperationException
+      case 1 =>
+        require(i == 0)
+        alleleRepr(c)
+      case 2 =>
+        require (i == 0 || i == 1)
+        val p = allelePair(c)
+        if (i == 0) p.j else p.k
+      case _ =>
+        require(i >= 0 && i < ploidy(c))
+        alleles(c)(i)
+    }
   }
 
-  def genArb: Gen[Call] =
-    for {
-      v <- Variant.gen
-      nAlleles = v.nAlleles
-      nGenotypes = triangle(nAlleles)
-      call <- Gen.choose(0, nGenotypes - 1)
-    } yield {
-      check(call, nAlleles)
-      call
+  def parse(s: String): Call = Parser.parseCall(s)
+
+  def toString(c: Call): String = {
+    val phased = isPhased(c)
+    val sep = if (phased) "|" else "/"
```

Comment:
can you use chars here, instead of strings?

Reply:
maybe it doesn't matter a huge amount.
This stuff could be a lot more efficient if it takes a StringBuilder, though.